packages/uploadthing/src/internal/handler.ts

```diff
@@ -305,230 +262,211 @@ const runRouteMiddleware = (opts: {
+      "InvalidURL",
+      (err) =>
+        new UploadThingError({
+          code: "INTERNAL_SERVER_ERROR",
+          message: err.message,
+        }),
+    ),
+  );
+  yield* Effect.logDebug(
+    "Retrieving presigned URLs from UploadThing. Callback URL is:",
+    callbackUrl.href,
+  );
 
-    yield* Effect.logDebug(
-      "Validating files meet the config requirements",
-      files,
-    );
-    const invalidConfigError = (e: { _tag: string; reason: string }) =>
-      new UploadThingError({
-        code: "BAD_REQUEST",
-        message: `Invalid config: ${e._tag}`,
-        cause: e.reason,
-      });
-    yield* assertFilesMeetConfig(files, parsedConfig).pipe(
-      Effect.catchTags({
-        FileSizeMismatch: invalidConfigError,
-        FileCountMismatch: invalidConfigError,
-        InvalidRouteConfig: invalidConfigError,
-        InvalidFileSize: invalidConfigError,
-        InvalidFileType: invalidConfigError,
-        UnknownFileType: invalidConfigError,
+  const presignedUrls = yield* fetchEffJson(
+    generateUploadThingURL("/api/prepareUpload"),
+    PresignedURLResponseSchema,
+    {
+      method: "POST",
+      body: JSON.stringify({
+        files: filesWithCustomIds,
+        routeConfig: parsedConfig,
+        metadata,
+        callbackUrl: callbackUrl.origin + callbackUrl.pathname,
+        callbackSlug: opts.slug,
       }),
-    );
+      headers: { "Content-Type": "application/json" },
+    },
+  );
 
-    const callbackUrl = yield* resolveCallbackUrl(opts).pipe(
-      Effect.tapError((error) =>
-        Effect.logError("Failed to resolve callback URL", error),
-      ),
-      Effect.catchTag(
-        "InvalidURL",
-        (err) =>
-          new UploadThingError({
-            code: "INTERNAL_SERVER_ERROR",
-            message: err.message,
-          }),
-      ),
-    );
-    yield* Effect.logDebug(
-      "Retrieving presigned URLs from UploadThing. Callback URL is:",
-      callbackUrl.href,
-    );
+  yield* Effect.logDebug("UploadThing responded with:", presignedUrls);
+  yield* Effect.logDebug("Sending presigned URLs to client");
 
-    const presignedUrls = yield* fetchEffJson(
-      generateUploadThingURL("/api/prepareUpload"),
-      PresignedURLResponseSchema,
-      {
-        method: "POST",
-        body: JSON.stringify({
-          files: filesWithCustomIds,
-          routeConfig: parsedConfig,
-          metadata,
-          callbackUrl: callbackUrl.origin + callbackUrl.pathname,
-          callbackSlug: opts.slug,
-        }),
-        headers: { "Content-Type": "application/json" },
-      },
+  let promise: Promise<unknown> | undefined = undefined;
+  if (opts.isDev) {
+    const layer = Layer.succeed(fetchContext, yield* fetchContext);
+    promise = Effect.forEach(
+      presignedUrls,
+      (file) => conditionalDevServer(file.key, opts.apiKey),
+      { concurrency: 10 },
+    ).pipe(
+      Effect.provide(ConsolaLogger),
+      Effect.provide(layer),
+      Effect.runPromise,
```

Comment:
Perhaps create a runtime with `ConsolaLogger`, and use that instead.
The fetchContextLayer would also be better as `Effect.provideService(fetchContext, yield* fetchContext)` as layers originally are optimised to construct once at startup (though there are exceptions)